File: WebClient SDK/src/default/ui/cardFiles/filesViewMetadata.ts
import {
  ViewMetadata,
  IViewMetadata,
  ViewColumnMetadata,
  ViewParameterMetadata
} from 'tessa/views/metadata';
import { LocalizationManager } from 'tessa/localization';
import { SortingColumn } from 'tessa/views';
import { SchemeType } from 'tessa/scheme';

export class FilesViewMetadata {
  static create(): IViewMetadata {
    const viewMetadata = new ViewMetadata();
    viewMetadata.alias = '__FilesView';
    viewMetadata.multiSelect = true;
    viewMetadata.enableAutoWidth = false;
    viewMetadata.defaultSortColumn = 'Caption';
    viewMetadata.defaultSortingColumns = [new SortingColumn('Caption')];

    FilesViewMetadata.addDefaultColumns(viewMetadata);
    FilesViewMetadata.addDefaultParameters(viewMetadata);

    return viewMetadata;
  }

  private static addDefaultColumns(viewMetadata: ViewMetadata) {
    const groupCaptionColumn = new ViewColumnMetadata();
    groupCaptionColumn.caption = LocalizationManager.instance.localize(
      '$UI_CardFiles_GroupCaptionField'
    );
    groupCaptionColumn.alias = 'GroupCaption';
    groupCaptionColumn.schemeType = SchemeType.NullableString;

    const categoryCaptionColumn = new ViewColumnMetadata();
    categoryCaptionColumn.caption = LocalizationManager.instance.localize(
      '$CardTypes_Controls_FileCategory'
    );
    categoryCaptionColumn.alias = 'CategoryCaption';
    categoryCaptionColumn.schemeType = SchemeType.NullableString;
    categoryCaptionColumn.disableGrouping = true;
    categoryCaptionColumn.localizable = true;

    const captionColumn = new ViewColumnMetadata();
    captionColumn.caption = LocalizationManager.instance.localize('$UI_CardFiles_CaptionField');
    captionColumn.alias = 'Caption';
    captionColumn.schemeType = SchemeType.String;
    captionColumn.sortBy = 'Caption';
    captionColumn.disableGrouping = true;
    // captionColumn.hasTag = true;

    const sizeAbsoluteColumn = new ViewColumnMetadata();
    sizeAbsoluteColumn.caption = 'SizeAbsolute';
    sizeAbsoluteColumn.alias = 'SizeAbsolute';
    sizeAbsoluteColumn.schemeType = SchemeType.UInt64;
    sizeAbsoluteColumn.sortBy = 'SizeAbsolute';
    sizeAbsoluteColumn.disableGrouping = true;
    sizeAbsoluteColumn.hidden = true;

    const sizeColumn = new ViewColumnMetadata();
    sizeColumn.caption = LocalizationManager.instance.localize('$UI_CardFiles_SizeField');
    sizeColumn.alias = 'Size';
    sizeColumn.schemeType = SchemeType.String;
    sizeColumn.sortBy = 'SizeAbsolute';
    sizeColumn.disableGrouping = true;

    viewMetadata.columns.set(groupCaptionColumn.alias, groupCaptionColumn);
    viewMetadata.columns.set(captionColumn.alias, captionColumn);
    viewMetadata.columns.set(categoryCaptionColumn.alias, categoryCaptionColumn);
    viewMetadata.columns.set(sizeAbsoluteColumn.alias, sizeAbsoluteColumn);
    viewMetadata.columns.set(sizeColumn.alias, sizeColumn);
  }

  private static addDefaultParameters(viewMetadata: ViewMetadata) {
    const nameParameter = new ViewParameterMetadata();
    nameParameter.alias = 'Caption';
    nameParameter.caption = LocalizationManager.instance.localize('$UI_CardFiles_CaptionField');
    nameParameter.schemeType = SchemeType.String;
    nameParameter.multiple = true;
    nameParameter.allowedOperands = ['Contains', 'StartWith', 'EndWith', 'Equality'];
    viewMetadata.parameters.set(nameParameter.alias, nameParameter);
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/CardFiles/FilesViewMetadata.cs
using Tessa.Extensions.Default.Shared.Views;
using Tessa.Localization;
using Tessa.Platform.Collections;
using Tessa.Properties.Resharper;
using Tessa.Scheme;
using Tessa.Views.Metadata;
using Tessa.Views.Metadata.Criteria;

namespace Tessa.Extensions.Default.Client.UI.CardFiles
{
    public static class FilesViewMetadata
    {
        /// <summary>
        ///     Создает метаданные для поставщика данных предоставляющего данные о файлах
        /// </summary>
        /// <returns>Метаданные</returns>
        [NotNull]
        public static IViewMetadata Create()
        {
            var viewMetadata = new ViewMetadata { Alias = "__FilesView", MultiSelect = true, EnableAutoWidth = true, DefaultSortColumn = ColumnsConst.Caption };
            AddDefaultColumns(viewMetadata);
            AddDefaultParameters(viewMetadata);
            return viewMetadata;
        }

        private static void AddDefaultParameters([NotNull] IViewMetadata viewMetadata)
        {
            var nameParameter = new ViewParameterMetadata
            {
                Alias = ColumnsConst.Caption,
                Caption = LocalizationManager.Localize(ColumnsConst.CaptionLocalization),
                SchemeType = SchemeType.String,
                Multiple = true,
                AllowedOperands = new[]
                    { CriteriaOperatorConst.Contains, CriteriaOperatorConst.StartWith, CriteriaOperatorConst.EndWith, CriteriaOperatorConst.Equality }
            };

            viewMetadata.Parameters.Add(nameParameter);

        }

        private static void AddDefaultColumns([NotNull] IViewMetadata viewMetadata)
        {
            var groupCaptionColumn = new ViewColumnMetadata
            {
                Caption = LocalizationManager.Localize(ColumnsConst.GroupLocalization),
                Alias = ColumnsConst.GroupCaption,
                SchemeType = SchemeType.NullableString
            };

            var categoryCaptionColumn = new ViewColumnMetadata
            {
                Caption = LocalizationManager.Localize(ColumnsConst.CategoryLocalization),
                Alias = ColumnsConst.CategoryCaption,
                SchemeType = SchemeType.NullableString,
                DisableGrouping = true,
                Localizable = true
            };

            var captionColumn = new ViewColumnMetadata
            {
                Caption = LocalizationManager.Localize(ColumnsConst.CaptionLocalization),
                Alias = ColumnsConst.Caption,
                SchemeType = SchemeType.String,
                SortBy = ColumnsConst.Caption,
                DisableGrouping = true,
                HasTag = true
            };

            var sizeAbsoluteColumn = new ViewColumnMetadata
            {
                Caption = ColumnsConst.SizeAbsolute,
                Alias = ColumnsConst.SizeAbsolute,
                SchemeType = SchemeType.UInt64,
                SortBy = ColumnsConst.SizeAbsolute,
                DisableGrouping = true,
                Hidden = true
            };

            var sizeColumn = new ViewColumnMetadata
            {
                Caption = LocalizationManager.Localize(ColumnsConst.SizeLocalization),
                Alias = ColumnsConst.Size,
                SchemeType = SchemeType.String,
                DisableGrouping = true,
                SortBy = ColumnsConst.SizeAbsolute,
            };

            viewMetadata.Columns.AddRange(
                groupCaptionColumn,
                captionColumn,
                categoryCaptionColumn,
                sizeAbsoluteColumn,
                sizeColumn
                );
        }
    }
}
